Feature/many to many support

pietaj opened the pull request:
I am targeting this branch, because the bundle does not support m:n auditing and the subject was already mentioned 
here https://github.com/sonata-project/EntityAuditBundle/issues/302, and here https://github.com/sonata-project/EntityAuditBundle/pull/321 but was never finished

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/EntityAuditBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Support for ManyToMany
```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [ ] Update the tests;
- [ ] Update the documentation;
- [ ] Add an upgrade note.
-->

halilibrahim committed:
Add audit tables for m:n relations

Code ported from https://github.com/halilibrahim/EntityAuditBundle/commit/862974f4b0be16af3f81dd47f12d48cfc8342b52
thx @halilibrahim§

VincentLanglet commented:
I fixed the cs in https://github.com/sonata-project/EntityAuditBundle/pull/510.
Please rebase and revert the extra style changes.

We don't work on this library a lot so the more the PR will be small, the easier it will be for us to review ^^

And take a look at the build ;)

Commit:
Add reading audited m:n relations

Code ported from https://github.com/halilibrahim/EntityAuditBundle/commit/474a29c69c14360102ec67ef97fbf6b76c1ec31e
thx @halilibrahim

Commit:
Add writing audited m:n relations

Code ported from https://github.com/halilibrahim/EntityAuditBundle/commit/20e703cbddc7a816df9e911fb1a32f3b985a73bb
thx @halilibrahim

Commit:
Add m:n tests

Commit:
fixup! Add writing audited m:n relations

pietaj commented:
I've rebased the code and reverted the style changes.

I've also left the deprecation fixes out of this PR as it makes the code harder to review.

pietaj commented:
@VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit?  What would be the most convenient way to proceed?

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
+                        $this->config->getRevisionTypeFieldName(),
+                    ];
+                    $tableName = $this->config->getTablePrefix(
+                        ).$assoc['joinTable']['name'].$this->config->getTableSuffix();
```
```suggestion
                    $tableName = $this->config->getTablePrefix()
                        .$assoc['joinTable']['name']
                        .$this->config->getTableSuffix();
```

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
+                            ', ',
+                            $columnList
+                        ).' FROM '.$tableName.' e WHERE '.$whereSQL.' ORDER BY e.'.$this->config->getRevisionFieldName(
+                        ).' DESC';
```
Prefer using sprintf

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
+                            $this->config->getRevisionTypeFieldName(),
+                        ];
+
+                        $tableName = $this->config->getTablePrefix(
+                            ).$targetAssoc['joinTable']['name'].$this->config->getTableSuffix();
```
```suggestion
                        $tableName = $this->config->getTablePrefix()
                            .$targetAssoc['joinTable']['name']
                            .$this->config->getTableSuffix();
```

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
+                        foreach ($targetAssoc['relationToSourceKeyColumns'] as $sourceKeyJoinColumn => $sourceKeyColumn) {
+                            $columnList[] = $sourceKeyJoinColumn;
+                        }
+                        $query = 'SELECT '.implode(
+                                ', ',
+                                $columnList
+                            ).' FROM '.$tableName.' e WHERE '.$whereSQL.' ORDER BY e.'.$this->config->getRevisionFieldName(
+                            ).' DESC';
```
prefer sprintf

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
+     * @return string
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
+     * @psalm-param array{cache?: array, cascade: array<string>, declared?: class-string, fetch: mixed, fieldName: string, id?: bool, inherited?: class-string, indexBy?: string, inversedBy: null|string, isCascadeRemove: bool, isCascadePersist: bool, isCascadeRefresh: bool, isCascadeMerge: bool, isCascadeDetach: bool, isOnDeleteCascade?: bool, isOwningSide: true, joinColumns?: array<array{name: st ...
```
You can use @phpstan-import-type AssociationMapping from ... https://github.com/doctrine/orm/blob/2.13.x/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L110
instead on the class level.

And then use it in the phpdoc.

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
     }
 
+    /**
+     * @param ClassMetadata $class
```
```suggestion
     * @param ClassMetadata<object> $class
```
avoid the need of psalm-param

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
+     *
+     * @return string
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
```
We prefer phpstan-param

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -403,34 +483,51 @@ private function saveRevisionEntityData(ClassMetadata $class, array $entityData,
             }
-            if (
-                ($assoc['type'] & ClassMetadata::TO_ONE) === 0
+            if (!(($assoc['type'] & ClassMetadata::TO_ONE) === 0
```
Change !($a || $b ||$c) to !a && !b && !c

VincentLanglet commented:
> @VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit? What would be the most convenient way to proceed?

Yes please fix the issues. 

Lot of them where already fix on some others places of the code with some if checks or some assert. You can do the same.

pietaj commented:
> > @VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit? What would be the most convenient way to proceed?
> 
> Yes please fix the issues.
> 
> Lot of them where already fix on some others places of the code with some if checks or some assert. You can do the same.

Oh my, thanks for the ultra fast reply, I didn't mean to push ;) I will fix all issues and will come back to you after everything will pass green

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

VincentLanglet commented:
Any issue @pietaj ? :)

Commit:
Style changes

Commit:
Psalm fixes

Commit:
PhpStan fixes

Commit:
More psalm fixes

pietaj commented:
My only issue was my day job :) did the requested psalm / phpstan compliance cleanup and the requested changes

VincentLanglet commented:
> My only issue was my day job :) did the requested psalm / phpstan compliance cleanup and the requested changes

Seems like build is still failing

pietaj commented:
That's odd ... both rector and phpstan pass on dev machine. I'll have a closer look

VincentLanglet commented:
> That's odd ... both rector and phpstan pass on dev machine. I'll have a closer look

Tru to update your dependencies maybe ?

pietaj commented:
yeah, thanks for the tip. Outdated dependencies, I'll rerun and fix the rest

Commit:
Use sprintf instead of string contat for queries

Commit:
Comply with cs-fixer rules

pietaj commented:
@VincentLanglet one of the failing checks is linked with rector, failing with

`PHP Fatal error:  Uncaught _PHPStan_9a6ded56a\Nette\DI\InvalidConfigurationException: The item 'conditionalTags › PHPStan\Rules\DisallowedConstructs\DisallowedLooseComparisonRule › phpstan.rules.rule' expects to be bool, array given. in phar:///home/runner/work/EntityAuditBundle/EntityAuditBundle/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php:219`

I did some quick checks and the error is linked with phpstan/phpstan bumping from `1.8.5 => 1.8.6`. Did you encounter a similar problems in other sonata projects as well?

jordisala1991 commented:
> @VincentLanglet one of the failing checks is linked with rector, failing with
> 
> `PHP Fatal error: Uncaught _PHPStan_9a6ded56a\Nette\DI\InvalidConfigurationException: The item 'conditionalTags › PHPStan\Rules\DisallowedConstructs\DisallowedLooseComparisonRule › phpstan.rules.rule' expects to be bool, array given. in phar:///home/runner/work/EntityAuditBundle/EntityAuditBundle/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php:219`
> 
> I did some quick checks and the error is linked with phpstan/phpstan bumping from `1.8.5 => 1.8.6`. Did you encounter a similar problems in other sonata projects as well?

Yes, ignore rector for now.

FYI, error already reported: https://github.com/rectorphp/rector/issues/7509, waiting for a fix. (The most you can do is upgrade to rector ^0.14, to ensure we will get the fix.

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -237,7 +237,9 @@ public function find($className, $id, $revision, array $options = [])
 
         foreach ($classMetadata->fieldNames as $columnName => $field) {
-            $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField($field) && !$classMetadata->isIdentifier($field)
+            $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField(
+                $field
+            ) && !$classMetadata->isIdentifier($field)
```
This reformatting is weird. I would prefer
```suggestion
            $tableAlias = $classMetadata->isInheritanceTypeJoined()
                && $classMetadata->isInheritedField($field)
                && !$classMetadata->isIdentifier($field)
```

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -404,7 +417,9 @@ public function findEntitiesChangedAtRevision($revision)
             foreach ($classMetadata->fieldNames as $columnName => $field) {
                 $type = Type::getType($classMetadata->fieldMappings[$field]['type']);
-                $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField($field) && !$classMetadata->isIdentifier($field)
+                $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField(
+                    $field
+                ) && !$classMetadata->isIdentifier($field)
```
```suggestion
                $tableAlias = $classMetadata->isInheritanceTypeJoined()
                    && $classMetadata->isInheritedField($field)
                    && !$classMetadata->isIdentifier($field)
```

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -541,7 +559,8 @@ public function findRevisions($className, $id)
 
         $query = 'SELECT r.* FROM '.$this->config->getRevisionTableName().' r '.
-                 'INNER JOIN '.$tableName.' e ON r.id = e.'.$this->config->getRevisionFieldName().' WHERE '.$whereSQL.' ORDER BY r.id DESC';
+            'INNER JOIN '.$tableName.' e ON r.id = e.'.$this->config->getRevisionFieldName(
+            ).' WHERE '.$whereSQL.' ORDER BY r.id DESC';
```
this formatting is weird

VincentLanglet commented on src/AuditReader.php:
```diff
@@ -889,7 +915,9 @@ private function createEntity($className, array $columnMap, array $data, $revisi
                             }
                         } elseif (null !== $mappedBy) {
-                            $otherEntityAssoc = $this->em->getClassMetadata($targetEntity)->associationMappings[$mappedBy];
+                            $otherEntityAssoc = $this->em->getClassMetadata(
+                                $targetEntity
+                            )->associationMappings[$mappedBy];
```
this formatting is weird

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -389,6 +426,45 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
+     * @param ClassMetadata        $class
+     * @param array<string, mixed> $assoc
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
```
```suggestion
     * @phpstan-param ClassMetadata<object> $targetClass
```

VincentLanglet commented on src/EventListener/LogRevisionsListener.php:
```diff
@@ -389,6 +426,45 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
+            && isset($assoc['relationToSourceKeyColumns'], $assoc['relationToTargetKeyColumns'], $assoc['joinTable']['name'])) {
+            $placeholders = ['?', '?'];
+
+            // $tableName = $this->config->getTableName($class);
```
Why adding a code commented ? can we remove it ?

VincentLanglet reviewed (changes requested):
It's ok for me, except some formatting issues.
Is this PR ok for you @jordisala1991 ?

jordisala1991 commented on tests/RelationTest.php:
```diff
@@ -306,8 +307,7 @@ public function testOneToOne(): void
     /**
-     * This test verifies the temporary behaviour of audited entities with M-M relationships
-     * until https://github.com/simplethings/EntityAudit/issues/85 is implemented.
+     * This test verifies the temporary behaviour of audited entities with M-M relationships.
```
I think this comment is outdated. The behavior implemented right now is not temporary right?

jordisala1991 reviewed (commented):
Do we need more information on the readme, or should we only remove the last todo?

jordisala1991 commented:
This Pr also needs a changelog.

Commit:
Bump rector to ^0.14

The version bump is necessary because phpstan/phpstan bump
1.8.5 => 1.8.6 breaks rector < 0.14.4
@see https://github.com/rectorphp/rector/issues/7509

Commit:
Cleanup code formatting

jordisala1991 commented:
Thanks @pietaj

jordisala1991 commented:
I guess the todo list on the readme should be updated too.
